flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/TimestampType.java

```diff
@@ -152,12 +152,14 @@ public boolean equals(Object o) {
 		if (!super.equals(o)) {
 			return false;
 		}
+
 		TimestampType that = (TimestampType) o;
-		return precision == that.precision;
+
+		return precision == that.precision && kind == that.kind;
```

Comment:
This was on purpose as mentioned in the comments of the constructor.

Reply:
Yes, `TimeIndicatorTypeInfo` and `SqlTimeTypeInfo.TIMESTAMP` also have same behavior.
But why are the kind does not affect equality and hashCode?
Is it for `Timestamp(6) of ROWTIME` and `Timestamp(6) of REGULAR` to interoperate?

Reply:
Yes, we don't want to have special path and checks for time attributes throughout the stack. If we want to check, we can use `getKind()` but the type fully behaves like a timestamp otherwise. Also, the serialized string representation is always just timestamp, so serialization and deserialization leads to objects that are `equal`.